qsr_lib/src/qsrlib_qstag/qstag.py

```diff
@@ -0,0 +1,515 @@
+                    source = temporal_names[self.graph.vs[source]['name']]
+                elif self.graph.vs[target]['node_type'] == 'temporal_relation':
+                    target = temporal_names[self.graph.vs[target]['name']]
+                dot_file.write('%s -> %s [arrowhead = "normal", color="red"];\n' %(source, target))
+        else:
+            for t_edge in self.temp_spatial_edges:
+                dot_file.write('%s -> %s [arrowhead = "normal", color="red"];\n' %(t_edge[0], t_edge[1]))
+
+        # Create spatial to object edges
+        for r_edge in self.spatial_obj_edges:
+            dot_file.write('%s -> %s [arrowhead = "normal", color="red"];\n' %(r_edge[0], r_edge[1]))
+        dot_file.write('}\n')
+        dot_file.close()
+
+
+def pretty_print_world_qsr_trace(which_qsr, qsrlib_response_message):
+    print(which_qsr, "request was made at ", str(qsrlib_response_message.req_made_at)
+          + " and received at " + str(qsrlib_response_message.req_received_at)
+          + " and finished at " + str(qsrlib_response_message.req_finished_at))
+    print("---")
+    print(qsrlib_response_message.qsrs.get_sorted_timestamps())
+    print("Response is:")
+    for t in qsrlib_response_message.qsrs.get_sorted_timestamps():
+        foo = str(t) + ": "
+        for k, v in zip(qsrlib_response_message.qsrs.trace[t].qsrs.keys(),
+                        qsrlib_response_message.qsrs.trace[t].qsrs.values()):
+            foo += str(k) + ":" + str(v.qsr) + "; "
+        print(foo)
+
+def compute_episodes(qsrlib_response_message):
+    """
+    Computes a long list of episodes with the format [(objects), {spatial relations}, (start, end)]
+    Or a dictionary where the objects are the key?
+    """
+
+    episodes  = []
+    #episodes  = {}
+    obj_based_qsr_world = {}
+    frames = qsrlib_response_message.qsrs.get_sorted_timestamps()
+    for frame in frames:
+        for objs, qsrs in qsrlib_response_message.qsrs.trace[frame].qsrs.items():
+            if objs not in obj_based_qsr_world: obj_based_qsr_world[objs] = []
+            obj_based_qsr_world[objs].append((frame, qsrs.qsr))
+
+    for objs, frame_tuples in obj_based_qsr_world.items():
+        #episodes[objs]  = []
+        epi_start, epi_rel = frame_tuples[0]
+        epi_end  = copy.copy(epi_start)
+
+        objects = objs.split(',')
+        for (frame, rel) in obj_based_qsr_world[objs]:
+            if rel == epi_rel:
+                epi_end = frame
+            else:
+                episodes.append( (objects, epi_rel, (epi_start, epi_end)) )
+                #episodes[objs].append( (objects, epi_rel, (epi_start, epi_end)) )
+
+                epi_start = epi_end = frame
+                epi_rel = rel
+        episodes.append( (objects, epi_rel, (epi_start, epi_end)) )
+        #episodes[objs].append( (objects, epi_rel, (epi_start, epi_end)) )
+    return episodes
+
+
+if __name__ == "__main__":
```

Comment:
Yes, by just looking the source. This file should not contain a `main` function, but create a `usage` script in the scripts folder.